core/headerchain.go

```diff
@@ -105,302 +122,212 @@ func NewHeaderChain(chainDb ethdb.Database, config *params.ChainConfig, engine c
 	}
 
 	hc.currentHeader.Store(hc.genesisHeader)
-	if head := rawdb.ReadHeadBlockHash(chainDb); head != (common.Hash{}) {
+	if head := rawdb.ReadHeadBlockHash(db); head != (common.Hash{}) {
 		if chead := hc.GetHeaderByHash(head); chead != nil {
 			hc.currentHeader.Store(chead)
 		}
 	}
 	hc.currentHeaderHash = hc.CurrentHeader().Hash()
 	headHeaderGauge.Update(hc.CurrentHeader().Number[types.QuaiNetworkContext].Int64())
 
+	//Initialize the heads slice
+	heads := make([]*types.Header, maxHeadsQueueLimit)
+	hc.heads = heads
+
+	//Load any state that is in our db
+	if err := hc.loadLastState(); err != nil {
+		return nil, err
+	}
+
 	return hc, nil
 }
 
-// GetBlockNumber retrieves the block number belonging to the given hash
-// from the cache or database
-func (hc *HeaderChain) GetBlockNumber(hash common.Hash) *uint64 {
-	if cached, ok := hc.numberCache.Get(hash); ok {
-		number := cached.(uint64)
-		return &number
+// Append
+func (hc *HeaderChain) Append(block *types.Block) error {
+	hc.headermu.Lock()
+	defer hc.headermu.Unlock()
+
+	// Append header to the headerchain
+	batch := hc.headerDb.NewBatch()
```

Comment:
We should attempt to append the block, before writing the header to the db. If the block fails to append (e.g. for invalid state transition or something), then that will throw an error and should prevent us from appending the header.